Add safety slot before variable for probe array

Godin opened the pull request:
Fixes #767

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeInserter.java:
```diff
@@ -62,7 +62,7 @@
 			pos += t.getSize();
 		}
-		variable = pos;
+		variable = pos + 1;
```
I would add some documentation about the "magic" +1. Either here and maybe in addition in the class documentation to explain the general approach.

Jacks0N23 commented:
Please merge it, it's block me af

AlexeyKorshun commented:
can you merge this fix?

Godin commented:
@Jacks0N23 @AlexeyKorshun 

> can you merge it?

unfortunately not yet - see https://github.com/jacoco/jacoco/pull/893#pullrequestreview-1403399584

Godin commented:
@marchof

> has probably no performance impact

I'm not so sure, are you? 😉 

> This will increase the byte code size

And method size is limited 😉 
 
> simply adding it to the end

I wouldn't call it "simply" - `ProbeInserter#visitMaxs` can not be used to determine position, because it is invoked after other `ProbeInserter` visit methods (such as `ProbeInserter#visitVarInsn`), which already should know position, so we will need to find another place/way to compute position. Which to me sounds more complex than a modification of this PR that will fix https://github.com/jacoco/jacoco/pull/893#discussion_r1178755535, and especially since I already have this modification and will publish it shortly 😉

Godin commented:
@marchof I rebased this PR and fixed the problem described in https://github.com/jacoco/jacoco/pull/893#discussion_r1178755535, it still lacks in-code comments requested in https://github.com/jacoco/jacoco/pull/893#discussion_r291450313 and the changelog entry, but maybe you already can review what else might be missing in the implementation?

----

Fun fact: the issue can be reproduced by using Kotlin compiler versions starting from released 7 years ago `1.0` 🙈 😆  up to the latest as of today released `1.8.21` - enough to declare an `inline` function with a default value for a parameter and body that immediately declares a variable of `Long` or `Double` type

```
inline fun example(x: Int = 0): Double {
    val y: Double = 0.0
    return x + y
}

fun main(args: Array<String>) {
}
```

marchof commented on org.jacoco.core.test/src/org/jacoco/core/internal/instr/SafetySlotTest.java:
```diff
@@ -0,0 +1,118 @@
+public class SafetySlotTest {
+
+	@Test
+	public void original() throws Exception {
```
As there is no assertion in the test maybe a descriptive test name can describe the purpose of the test:

```suggestion
	public void should_verify_original_class_without_errors() throws Exception {
```

Godin replied:
```suggestion
	public void jvm_should_verify_original_class_without_errors() throws Exception {
```

marchof commented on org.jacoco.core.test/src/org/jacoco/core/internal/instr/SafetySlotTest.java:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	@Test
+	public void instrumented() throws Exception {
```
```suggestion
	public void should_verify_instrumented_class_without_errors() throws Exception {
```

Godin replied:
```suggestion
	public void jvm_should_verify_instrumented_class_without_errors() throws Exception {
```

Godin commented:
@marchof I added comments and updated changelog, so IMO this can be merged 🎉
